File: Genetic_Algorithm-3/FFNNBinaryLatih.py
import numpy as np
import matplotlib.pyplot as plt
import json
import random

from BangMatrixIT import BangMatrixIT
from InisialisasiPopulasi import inisialisasi_populasi
from DekodekanKromosom import dekodekan_kromosom
from BinaryEvalInd import BinaryEvalInd
from LinearFitnessRanking import linear_fitness_ranking
from RouletteWheel import roulette_wheel
from PindahSilang import pindah_silang
from Mutasi import mutasi


# ==============================
# PARAMETER GA
# ==============================
JumMasukan = 3
JPmasukan = 2**JumMasukan
Nbit = 20
JumGen = Nbit * (JumMasukan + 1)**2
Nvar = int(JumGen / Nbit)

Rb = -10
Ra = 10
MinDelta = 0.01
Fthreshold = 1 / MinDelta
Bgraf = Fthreshold

UkPop = 100
Psilang = 0.8
Pmutasi = 0.03
MaxG = 2000


# ==============================
# GRAFIK
# ==============================
plt.ion()
fig, ax = plt.subplots(figsize=(6, 4))
fig.canvas.manager.set_window_title("Algoritma Genetika")

ax.set_title("Algoritma Genetika untuk Pelatihan FFNN (Binary Encoding)")
ax.set_xlim(1, MaxG)
ax.set_ylim(0, Bgraf)

best_arr = np.zeros(MaxG)
avg_arr = np.zeros(MaxG)

hbest, = ax.plot(range(1, MaxG+1), best_arr, label="Fitness Terbaik")
havg, = ax.plot(range(1, MaxG+1), avg_arr, label="Fitness Rata-rata")

ax.set_xlabel("Generasi")
ax.set_ylabel("Fitness")
ax.legend()
plt.draw()
plt.pause(0.01)


# ==============================
# DATASET (IM & TM)
# ==============================
IM, TM = BangMatrixIT(JumMasukan, JPmasukan)


# ==============================
# POPULASI AWAL
# ==============================
Populasi = inisialisasi_populasi(UkPop, JumGen)


# ==============================
# VARIABEL BEST SEPANJANG SEJARAH
# ==============================
best_global_fitness = -999999999
best_global_chrom = None


# ==============================
# LOOP GA
# ==============================
for generasi in range(1, MaxG + 1):

    fitness = np.zeros(UkPop)

    # ========== HITUNG FITNESS ==========
    for i in range(UkPop):
        FFNNstruk = dekodekan_kromosom(Populasi[i], Nvar, Nbit, Ra, Rb)
        fitness[i] = BinaryEvalInd(FFNNstruk, JumMasukan, JPmasukan, IM, TM)

    # Info generasi
    MaxF = np.max(fitness)
    MinF = np.min(fitness)
    idx_best = np.argmax(fitness)

    print(f"Generasi {generasi}: MaxF={MaxF:.5f}, MinF={MinF:.5f}")

    # Simpan best global sepanjang sejarah
    if MaxF > best_global_fitness:
        best_global_fitness = MaxF
        best_global_chrom = Populasi[idx_best].copy()

    # ========== UPDATE GRAFIK ==========
    best_arr[generasi-1] = MaxF
    avg_arr[generasi-1] = np.mean(fitness)

    hbest.set_ydata(best_arr)
    havg.set_ydata(avg_arr)
    plt.pause(0.01)

    # ========== STOP JIKA FIT THRESHOLD TERCAPAI ==========
    if MaxF >= Fthreshold:
        print(">>> Threshold tercapai. GA berhenti.")
        break

    # =======================
    # SELEKSI (RANKING + ROULETTE)
    # =======================
    ranking = linear_fitness_ranking(UkPop, fitness, MaxF, MinF)
    TempPop = Populasi.copy()

    # Elitisme
    TempPop[0] = Populasi[idx_best].copy()

    # =======================
    # PROSES GENERASI BARU
    # =======================
    for j in range(1, UkPop, 2):

        # Roulette Wheel memilih dua induk
        p1 = roulette_wheel(UkPop, ranking)
        p2 = roulette_wheel(UkPop, ranking)

        parent1 = Populasi[p1]
        parent2 = Populasi[p2]

        # Crossover
        if np.random.rand() <= Psilang:
            anak1, anak2 = pindah_silang(parent1, parent2)
        else:
            anak1 = parent1.copy()
            anak2 = parent2.copy()

        # Mutasi
        anak1 = mutasi(anak1, Pmutasi)
        anak2 = mutasi(anak2, Pmutasi)

        # Masukkan ke populasi baru
        TempPop[j] = anak1
        if j+1 < UkPop:
            TempPop[j+1] = anak2

    Populasi = TempPop.copy()

# END FOR


# ==============================
# SIMPAN MODEL TERBAIK
# ==============================
print("\nMenyimpan model terbaik...")

def convert_to_jsonable(obj):
    if isinstance(obj, np.ndarray):
        return obj.tolist()
    if isinstance(obj, (np.integer, np.floating)):
        return obj.item()
    if isinstance(obj, dict):
        return {k: convert_to_jsonable(v) for k, v in obj.items()}
    if isinstance(obj, list):
        return [convert_to_jsonable(v) for v in obj]
    return obj


best_model = dekodekan_kromosom(best_global_chrom, Nvar, Nbit, Ra, Rb)

with open("FFNN_terbaik.json", "w") as f:
    json.dump(convert_to_jsonable(best_model), f, indent=4)

print("Model terbaik disimpan ke FFNN_terbaik.json")

plt.ioff()
plt.show()

File: TA/FFNNBinaryLatih.m
clear all

JumMasukan = 3;
JPmasukan = 2^JumMasukan;
Nbit = 20;
JumGen = Nbit*(JumMasukan+1)^2;
Nvar = JumGen/Nbit;
Rb = -10;
Ra = 10;
MinDelta = 0.01;
Fthreshold = 1/MinDelta;
Bgraf = Fthreshold;

UkPop = 40;
Psilang = 0.8;
Pmutasi = 0.03;
MaxG = 2000;

%inisialisasi Grafis
hfig = figure;
hold on
title('Algortima Genetika dengan BInary Encoding untuk pelatihan FFNN');
set(hfig, 'position', [50,50,600,400]);
set(hfig,'DoubleBuffer', 'on');
axis([1 MaxG 0 Bgraf]);
hbestplot1 = plot(1:MaxG,zeros(1,MaxG));
hbestplot2 = plot(1:MaxG,zeros(1,MaxG));
htext1 = text(0.6*MaxG,0.25*Bgraf,sprintf('Fitness terbaik: %7.6f', 0.0));
htext2 = text(0.6*MaxG,0.20*Bgraf,sprintf('Fitness rata-rata: %7.6f', 0.0));
htext3 = text(0.6*MaxG,0.15*Bgraf,sprintf('Ukuran Populasi: %3.0f', 0.0));
htext4 = text(0.6*MaxG,0.10*Bgraf,sprintf('Prob. Pindah Silang: %4.3f', 0.0));
htext5 = text(0.6*MaxG,0.05*Bgraf,sprintf('Prob. Mutasi: %4.3f', 0.0));
xlabel('Generasi');
ylabel('Fitness');
hold off
drawnow;

[IM,TM] = BangMatrixIT(JumMasukan,JPmasukan);

Populasi  = InisialisasiPopulasi(UkPop,JumGen);

for generasi = 1:MaxG
    FFNNstruk = DekodekanKromosom(Populasi(1,:),Nvar,Nbit,Ra,Rb);
    Fitness(1) = BinaryEvalInd(FFNNstruk, JumMasukan, JPmasukan,IM,TM);
    MaxF = Fitness(1);
    MinF = Fitness(1);
    IndeksIndividuTerbaik = 1;
    for ii=2:UkPop
        FFNNstruk = DekodekanKromosom(Populasi(ii,:),Nvar,Nbit,Ra,Rb);
        Fitness(ii) = BinaryEvalInd(FFNNstruk, JumMasukan, JPmasukan,IM,TM);
        if (Fitness(ii) > MaxF)
            MaxF = Fitness(ii);
            IndeksIndividuTerbaik = ii;
            FFNNterbaik = FFNNstruk;
        end
        if (Fitness(ii) <= MinF)
            MinF = Fitness(ii);
        end
    end

    FitnessRataRata = mean(Fitness);

    plotvector1 = get(hbestplot1,'YData');
    plotvector1(generasi) = MaxF;
    set(hbestplot1,'YData', plotvector1);
    plotvector2 = get(hbestplot2,'YData');
    plotvector2(generasi) = FitnessRataRata;
    set(hbestplot2,'YData', plotvector2);
    set(htext1,'String',sprintf('Fitness terbaik: %7.6f', MaxF));
    set(htext2,'String',sprintf('Fitness rata-rata: %7.6f', FitnessRataRata));
    set(htext3,'String',sprintf('Ukuran Populasi: %3.0f', UkPop));
    set(htext4,'String',sprintf('Prob. Pindah Silang: %4.3f', Psilang));
    set(htext5,'String',sprintf('Prob. Mutasi: %4.3f', Pmutasi));
    drawnow

    if MaxF > Fthreshold
        break;
    end

    TemPopulasi = Populasi;

    if mod(UkPop,2) == 0
        IterasiMulai = 3;
        TemPopulasi(1,:) = Populasi(IndeksIndividuTerbaik,:);
        TemPopulasi(2,:) = Populasi(IndeksIndividuTerbaik,:);
    else
        IterasiMulai = 2;
        TemPopulasi(1,:) = Populasi(IndeksIndividuTerbaik,:);
    end

    for jj = IterasiMulai:2:UkPop
        IP1 = RankSelection(UkPop);
        IP2 = RankSelection(UkPop);

        if(rand < Psilang)
            Anak = PindahSilang(Populasi(IP1,:), Populasi(IP2,:), JumGen);
            TemPopulasi(jj,:) = Anak(1,:);
            TemPopulasi(jj+1,:) = Anak(2,:);
        else
            TemPopulasi(jj,:) = Populasi(IP1,:);
            TemPopulasi(jj+1,:) = Populasi(IP2,:);
        end
    end

    for kk = IterasiMulai:UkPop
        TemPopulasi(kk,:) = MutasiRandomReset(TemPopulasi(kk,:), JumGen,Pmutasi);
    end

    Populasi = TemPopulasi;

end

FFNNterbaik;
generasi
FitnessRataRata
fitness_terbaik = max(Fitness)
save FFNNterbaik.mat FFNNterbaik

        

